File: 146.lru-cache.java
class LRUCache {

    private Map<Integer, Node> m;
    private Node head;
    private Node tail;
    private int capacity;

    public LRUCache(int capacity) {
        this.m = new HashMap<Integer, Node>();
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);
        this.head.next = tail;
        this.tail.prev = head;
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if (!this.m.containsKey(key)) {
            return -1;
        }
        Node n = this.m.get(key);
        remove(n);
        appendToTail(n);
        return n.value;
    }
    
    public void put(int key, int value) {
        if (this.capacity < 1) {
            return;
        }
        if (this.get(key) != -1) {
            this.m.get(key).value = value;
            return;
        }
        if (this.capacity == this.m.size()) {
            this.m.remove(this.head.next.key);
            remove(this.head.next);
        }
        Node n = new Node(key, value);
        this.m.put(key, n);
        appendToTail(n);
    }

    private void appendToTail(Node n) {
        this.tail.prev.next = n;
        n.prev = this.tail.prev;
        this.tail.prev = n;
        n.next = this.tail;
    }

    private void remove(Node n) {
        n.prev.next = n.next;
        n.next.prev = n.prev;
        n.prev = null;
        n.next = null;
    }
}

class Node {
    int key;
    int value;
    Node prev;
    Node next;

    Node(int key, int value) {
        this.key = key;
        this.value = value;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146.lru-cache.go
type LRUCache struct {
	cache      map[int]*Node
	head, tail *Node
	capacity   int
}

type Node struct {
	key, value int
	prev, next *Node
}

func Constructor(capacity int) LRUCache {
	head, tail := &Node{}, &Node{}
	head.next = tail
	tail.prev = head
	return LRUCache{make(map[int]*Node), head, tail, capacity}
}

func (this *LRUCache) Get(key int) int {
	n, ok := this.cache[key]
	if !ok {
		return -1
	}
	this.remove(n)
	this.insert(n)
	return n.value
}

func (this *LRUCache) Put(key int, value int) {
	if this.Get(key) != -1 {
		this.cache[key].value = value
		return
	}
	if len(this.cache) == this.capacity {
		n := this.head.next
		delete(this.cache, n.key)
		this.remove(n)
	}
	n := &Node{key: key, value: value}
	this.cache[key] = n
	this.insert(n)
}

func (this *LRUCache) remove(n *Node) {
	n.prev.next = n.next
	n.next.prev = n.prev
	n.prev, n.next = nil, nil
}

func (this *LRUCache) insert(n *Node) {
	this.tail.prev.next = n
	n.prev = this.tail.prev
	this.tail.prev = n
	n.next = this.tail
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
